crates/opactl/src/main.rs

```diff
@@ -583,11 +622,127 @@ pub mod test {
+        context: &mut TestTransactionContext,
+        handler: &OpaTransactionHandler,
+    ) -> (Vec<(String, Value)>, PrintableEvent) {
+        apply_transactions(handler, context, transactions);
+        (context.readable_state(), context.readable_events())
+    }
+
+    fn test_determinism(
+        transactions: &[sawtooth_sdk::messages::transaction::Transaction],
+        context: &Arc<Mutex<TestTransactionContext>>,
+        number_of_determinism_checking_cycles: usize,
+    ) -> Vec<(Vec<(String, Value)>, PrintableEvent)> {
+        let handler = OpaTransactionHandler::new();
+
+        let contexts = (0..number_of_determinism_checking_cycles)
+            .into_par_iter()
+            .map(|_| {
+                let mut context = context.lock().unwrap().clone();
+                process_transactions(transactions, &mut context, &handler)
+            })
+            .collect::<Vec<_>>();
+
+        // Check if the contexts are the same after running apply
+        assert!(
+            contexts.par_iter().all(|context| contexts[0] == *context),
+            "All contexts must be the same after running apply. Contexts: {:?}",
+            contexts,
+        );
+
+        contexts
+    }
+
+    fn assert_output_determinism(
+        expected_contexts: &[(Vec<(String, Value)>, PrintableEvent)],
+        readable_state_and_events: &(Vec<(String, Value)>, PrintableEvent),
+    ) {
+        // Check if the updated context is the same as the determinism check results
+        assert!(
+            expected_contexts
+                .iter()
+                .all(|context| readable_state_and_events == context),
+            "Updated context must be the same as previously run tests"
+        );
+    }
+
+    #[derive(Clone)]
     struct WellBehavedBehavior {
         handler: Arc<OpaTransactionHandler>,
         context: Arc<Mutex<TestTransactionContext>>,
     }
 
+    impl WellBehavedBehavior {
+        /// Submits a batch of transactions to the validator and performs determinism checks.
+        fn submit_batch(&self, request: &[u8]) -> Result<Vec<u8>, SawtoothCommunicationError> {
+            // Parse the request into a `ClientBatchSubmitRequest` object and extract the first batch.
+            let mut req = ClientBatchSubmitRequest::parse_from_bytes(request).unwrap();
+            let mut batch = req.take_batches().into_iter().next().unwrap();
+
+            // Log some debug information about the batch and sort its transactions.
+            debug!(received_batch = ?batch, transactions = ?batch.transactions);
+            let transactions = get_sorted_transactions(&mut batch);
+
+            // Perform a determinism checking cycle for a specified number of times
+            let number_of_determinism_checking_cycles = 5;
+
+            // Use the `test_determinism` function to perform determinism checking and get the expected contexts
+            let context = Arc::clone(&self.context);
```

Comment:
This is going to guarantee the sam context, as that arc contains a reference to the same context.

We need a deep / non ref copy or a way of constructing a default one here. A constraint on context: default::Default might do it - that way we can make new empty ones